numba/types/functions.py

```diff
@@ -43,23 +105,30 @@ def get_impl_key(self, sig):
         return self._impl_keys[sig.args]
 
     def get_call_type(self, context, args, kws):
-        for temp_cls in self.templates:
-            temp = temp_cls(context)
-            sig = temp.apply(args, kws)
-            if sig is not None:
-                self._impl_keys[sig.args] = temp.get_impl_key(sig)
-                return sig
+        return self.get_call_type_with_literals(context, args, kws,
+                                                literals=None)
 
     def get_call_type_with_literals(self, context, args, kws, literals):
+        failures = _ResolutionFailures(context, self, args, kws)
         for temp_cls in self.templates:
             temp = temp_cls(context)
-            if literals is not None and temp.support_literals:
-                sig = temp.apply(*literals)
+            try:
+                if literals is not None and temp.support_literals:
+                    sig = temp.apply(*literals)
+                else:
+                    sig = temp.apply(args, kws)
+            except Exception as e:
+                sig = None
+                failures.add_error(temp_cls, e)
             else:
-                sig = temp.apply(args, kws)
-            if sig is not None:
-                self._impl_keys[sig.args] = temp.get_impl_key(sig)
-                return sig
+                if sig is not None:
+                    self._impl_keys[sig.args] = temp.get_impl_key(sig)
+                    return sig
+                else:
+                    failures.add_error(temp_cls, "rejected")
+
+        if failures:
```

Comment:
Is this `if` needed @sklam ?

Reply:
You're right.  I will turn it into a check for the unexpected case `bool(failures)==False`